hnn_core/tests/test_extracellular.py

```diff
@@ -0,0 +1,341 @@
+        _ = rec_arr.sfreq
+
+
+def test_transmembrane_currents():
+    """Test that net transmembrane current is zero at all times."""
+    hnn_core_root = op.dirname(hnn_core.__file__)
+    params_fname = op.join(hnn_core_root, 'param', 'default.json')
+    params = read_params(params_fname)
+    params.update({'N_pyr_x': 3,
+                   'N_pyr_y': 3,
+                   'tstop': 40,
+                   't_evprox_1': 5,
+                   't_evdist_1': 10,
+                   't_evprox_2': 20,
+                   'N_trials': 1})
+    net = default_network(params, add_drives_from_params=True)
+    electrode_pos = (0, 0, 0)  # irrelevant where electrode is
+    # all transfer resistances set to unity
+    net.add_electrode_array('net_Im', electrode_pos, method=None)
+    _ = simulate_dipole(net, postproc=False)
+    currents = net.rec_array['net_Im'].get_data()
+    assert_allclose(currents, 0, rtol=1e-10, atol=1e-10)
+
+
+def test_transfer_resistance():
+    """Test transfer resistances calculated correctly"""
+    from neuron import h
+    from hnn_core.extracellular import _transfer_resistance
+    sec = h.Section(name='dend')
+    h.pt3dclear(sec=sec)
+    h.pt3dadd(0, 0, 0, 1, sec=sec)
+    h.pt3dadd(0, 1, 0, 1, sec=sec)  # section oriented along y-axis
+    sec.L = 300
+    sec.diam = 8
+    sec.nseg = 5
+    # NB segment lengths aren't equal! First/last segment center point is
+    # closer to respective end point than to next/previous segment!
+    seg_ctr_pts = [0]
+    seg_ctr_pts.extend([seg.x * sec.L for seg in sec])
+    seg_ctr_pts.append(sec.L)
+    seg_lens = np.diff(seg_ctr_pts)
+    first_len = seg_lens[0]
+    seg_lens = np.array([first_len] + list(seg_lens[2:]))
+    seg_ctr_pts = seg_ctr_pts[1:-1]  # remove end points again
+
+    conductivity = 0.3
+
+    elec_pos = (10, 150, 0)
+    target_vals = {'psa': list(), 'lsa': list()}
+    for seg_idx in range(sec.nseg):
+        # PSA: distance to middle segment == electrode x-position
+        var_r_psa = np.sqrt(elec_pos[0] ** 2 +
+                            (elec_pos[1] - seg_ctr_pts[seg_idx]) ** 2)
+        target_vals['psa'].append(
+            1000 / (4. * np.pi * conductivity * var_r_psa))
+
+        # LSA: calculate L and H variables relative to segment endpoints
+        var_l = elec_pos[1] - (seg_ctr_pts[seg_idx] - seg_lens[seg_idx])
+        var_h = elec_pos[1] - (seg_ctr_pts[seg_idx] + seg_lens[seg_idx])
+        var_r_lsa = elec_pos[0]  # just use the axial distance
+        target_vals['lsa'].append(
+            1000 * np.log(np.abs(
+                (np.sqrt(var_h ** 2 + var_r_lsa ** 2) - var_h) /
+                (np.sqrt(var_l ** 2 + var_r_lsa ** 2) - var_l)
+            )) / (4. * np.pi * conductivity * 2 * seg_lens[seg_idx]))
+
+    for method in ['psa', 'lsa']:
+        res = _transfer_resistance(sec, elec_pos, conductivity, method)
+        assert_allclose(res, target_vals[method])
```

Comment:
why `allclose` and not an exact equal?

Reply:
Good question. I guess I just avoid "exactness" when it comes to floating point calculations. Note that the test is an independent implementation of the maths in the calculating function.

I've reduced `rtol` to `1e-12`, doesn't get much more 'exact' than that :)